### DISCUSSION TITLE
Rework "Supported-Devices" Page

### DISCUSSION THREAD

**Discussion by psi-4ward** [Koenkk/zigbee2mqtt#8928] [General]
Hi mates,

I thought about reworking the docu page to use something more flexible than jekyll. 

For https://asksinpp.de I've used VitePress with algolia search plugin but especially for the _supported devices_ page we could introduce some more advanced navigation/filtering instead of this growing loooooong list. As simple example something like https://asksinpp.de/Projekte/ but perhaps with more granular filters like Vendor, Type (Switch, Sensor, Bulb) or other attributes. 

VitePress is somewhat similar to Jekyll and generates a static page from Markdown but with the power of Vue.js to do some nifty things with VueCompoents (like interactive filtering).

It should not to much work to switch, most work would be the (optional?) adding of attributes to the devices for the filtering. 

What do you think? Is it worth to investigate? (In my op, it would be nice to be able to filter the supported devices for given usecases)

**Comment by Koenkk** [comment]
I believe such filters are also possible with Jekyll, I would prefer not to switch for just this. Jekyll is very well supported by Github and more popular.

**Comment by psi-4ward** [comment]
Just a little first draft of a possible devices-overview page:
https://tmp.psi.cx/z/

What do you think? And of course we can integrate this in jekyll.

**Reply by Koenkk** [reply]
Looks good! However could you make the page less wide? All other pages are also not full width.

**Reply by psi-4ward** [reply]
Glad you like it. And of course, it's a draft and the max-width is easy to adjust. 
I'll go and investigate some more and prepare a PR.

**Reply by psi-4ward** [reply]
Just deployed some more improvements. Pls check it out ;)

**Reply by Koenkk** [reply]
Looks good!

Looks like some layout issue when the device description is > 1 line:
<img width="625" alt="Screenshot 2021-10-12 at 19 32 25" src="https://user-images.githubusercontent.com/2892853/137002711-8e56f64d-0de4-435f-9aef-5ade64c154ea.png">

**Reply by psi-4ward** [reply]
Which browser do you use?

**Reply by Koenkk** [reply]
Chrome

**Reply by psi-4ward** [reply]
Fixed :)

**Reply by psi-4ward** [reply]
I've integrated it into the jekyll build, you can review the result at https://tmp.psi.cx/z/

It should also work on smaller devices.

**Comment by psi-4ward** [comment]
I didn't found any github actions to build the page. 
Looks like Jekyll gets invoked by github-pages and you docman manually to update the generated pages?

The devices-overview app is built Vue.js so I would invoke it in `ci.yml`?

**Reply by psi-4ward** [reply]
Deployed it to https://psi-4ward.github.io/zigbee2mqtt.io/information/supported_devices.html and added the vue-app prod-build assets to the repo.

**Reply by Koenkk** [reply]
Yes you are right, I check in the generated pages manually. Maybe we can generate the supported device page with the `npm run docgen` command? https://github.com/Koenkk/zigbee2mqtt.io/blob/554a91df6056af8baba4d69ba70cc7c8724eeda2/package.json#L21

Really like the looks of this page!

**Reply by psi-4ward** [reply]
It is already implemented using docgen: https://github.com/psi-4ward/zigbee2mqtt.io/blob/master/docgen/supported-devices.js

This page bootstraps the app and holds the list of shuffled deviced. I decided to shuffle that not always the same devices stay on top.

I kept the table-view also: https://github.com/psi-4ward/zigbee2mqtt.io/blob/master/docgen/supported-devices-table.js

To build app rendering the page we could manually do:
```bash
cd devices-overview
npm ci && npm run build
cd ..
git add docs/assets
```

But I think the vue-app should not change very often ;)

If you're happy with this I'll create the PR.

PS:
* A discord-user mentioned that it makes more sense to use an AND condition for the exposes filter - I've changed the behavior
* I've added RegEx support for text-search. Most users won't use it but it seems to perform good enough.
* If one selects any filter the devices gets sorted by vendor+model
* The search terms getting encoded in the URL so one can send "Supported Bulbs from IKEA" set to a friend.
* Cause of the lazy-loading behavior, I store the count of loaded pages and the scroll-offset for the current session to make browser-back work.

**Reply by Koenkk** [reply]
Thanks, looking forward to the PR

**Reply by psi-4ward** [reply]
At reason from some discussion in discord I now read the added-date from git (when the device.md was initial added) and use it for sorting and label new devices (added not older 30d).

Cause it's a bit IO heavy I've rewritten docgen a bit to be able to execute tasks in parallel.

**Comment by der-pw** [comment]
Is it possible that by clicking on the name, the filter for vendor is executed immediately?
The same I could imagine for the exposes.

![image](https://user-images.githubusercontent.com/44023915/137453242-bf2834cc-0b82-4980-8f85-a50100f04e6d.png)

**Reply by psi-4ward** [reply]
Of course. Let's see what @Koenkk thinks :-)

**Reply by Koenkk** [reply]
Good idea!

**Reply by psi-4ward** [reply]
Should it also work for clicking a expose?

**Comment by kirovilya** [comment]
Can we add comments and ratings later? :)
Still need a classification of devices: switch, plug, bulb, button, sensor...

**Reply by psi-4ward** [reply]
Difficult. We need a sane storage and some type of authentication. Probably you would need an github.com account for that

**Reply by kirovilya** [reply]
https://aristath.github.io/blog/static-site-comments-using-github-issues-api
comments - not that hard, but it will allow you to post reviews

**Reply by kirovilya** [reply]
sorry, didn’t notice that it’s not a jekyll page now

**Reply by psi-4ward** [reply]
so comments on the overview page does not make much sense. they should go into device-detail pages to be bound to a specific device.

ratings are another problem cause of something like a database and the github requests are rate-limited.
